File: scripts/minfont.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

"""
minfont.js 的 Python 版，速度没有 minfont.js 的快，但生成的字体文件大小更小
"""

import os
import platform
from os.path import basename, exists, join

from fontTools import subset
from fontTools.ttLib import TTFont

origin_path = join('docs')
ignore_dirs = ['dist', 'public', '.DS_Store']
out_path = join('docs', 'public', 'fonts')
fonts = [
    join('fonts', 'HarmonyOS_Sans_SC_Bold.ttf'),
    join('fonts', 'HarmonyOS_Sans_SC_Regular.ttf'),
]

content = ''


def read(path):
    global content
    for root, dirs, files in os.walk(path):
        if basename(root) in ignore_dirs:
            dirs[:] = []  # 忽略当前目录下的子目录
            continue
        for f in files:
            print(f'正在读取 {join(root, f)}')
            with open(join(root, f), 'r', encoding='utf8') as file:
                while True:
                    if char := file.read(1):
                        content += f'{char}\n'
                    else:
                        break


print('读取文件中...')
read(origin_path)

for font in fonts:
    print(f'正在处理 {font}')
    if not exists(out_path):
        os.mkdir(out_path)
    f = TTFont(font)

    subsetter = subset.Subsetter()

    subsetter.populate(text=content)
    subsetter.subset(f)
    f.flavor = 'woff2'
    f.save(join(out_path, font.replace(".ttf", ".woff2")))

File: scripts/minfont.js
/*
 * minfont.py 的 JS 版，速度比 minfont.py 快，但生成的字体文件大小更大
 */

const fs = require('fs') // 不可转为 import
const Fontmin = require('fontmin') // 不可转为 import

const originPath = 'docs/'
const ignoreDirs = ['dist', 'public', '.DS_Store']
const outPath = 'docs/public/fonts/'
const fonts = [
  'fonts/HarmonyOS_Sans_SC_Bold.ttf',
  'fonts/HarmonyOS_Sans_SC_Regular.ttf',
]

let content = ''

function readFile(file) {
  const stat = fs.statSync(`${originPath}${file}`)
  if (stat.isFile()) {
    console.log(`正在读取 ${originPath}${file}`)
    const md = fs.readFileSync(`${originPath}${file}`)
    content += md.toString()
  }
  if (stat.isDirectory()) {
    const subdir = fs.readdirSync(`${originPath}${file}`)
    for (let i = 0; i < subdir.length; i++) {
      if (ignoreDirs.includes(subdir[i])) continue
      readFile(`${file}/${subdir[i]}`)
    }
  }
}

function minFont(font) {
  console.log(`正在处理 ${font}`)
  const fontmin = new Fontmin()
    .src(font)
    .use(Fontmin.glyph({ text: content }))
    .use(Fontmin.ttf2woff2())
    .dest(outPath)

  fontmin.run(function (err) {
    if (err) {
      console.log(`处理 ${font} 过程中出现错误：`)
      console.error(err)
    }
  })
}

console.log('读取文件中...')
fs.readdir(originPath, (err, files) => {
  if (err) {
    return console.error(err)
  }
  for (let i = 0; i < files.length; i++) {
    if (ignoreDirs.includes(files[i])) continue
    readFile(files[i])
  }
  if (!fs.existsSync(outPath)) fs.mkdirSync(outPath)
  for (const idx in fonts) {
    minFont(fonts[idx])
  }
})
